src/forge/api/types.py

```diff
@@ -0,0 +1,179 @@
+            Values can be scalars, tensors, or other structured data depending on the loss.
+
+    Example:
+        >>> result = await trainer.forward_backward(batch)
+        >>> print(f"Loss: {result.loss:.4f}")
+        >>> if result.metrics:
+        >>>     print(f"Metrics: {result.metrics}")
+    """
+
+    loss: float
+    metrics: dict[str, Any]
+
+
+@dataclass
+class OptimStepResult:
+    """Result from an optimizer step.
+
+    Attributes:
+        step: Training step number after this optimizer step
+        learning_rate: Current learning rate used for this step
+        accumulated_microbatches: Number of forward_backward calls that were
+            accumulated before this optimizer step. Useful for tracking gradient
+            accumulation behavior.
+
+    Example:
+        >>> result = await trainer.optim_step()
+        >>> print(f"Step {result.step}, LR={result.learning_rate:.2e}")
+        >>> print(f"Accumulated {result.accumulated_microbatches} batches")
+    """
+
+    step: int
+    learning_rate: float
+    accumulated_microbatches: int
+
+
+@dataclass
+class ParallelismInfo:
+    """Parallelism configuration for distributed training.
+
+    Attributes:
+        dp_degree: Data parallel degree (number of data parallel replicas)
+        tp_degree: Tensor parallel degree (model sharding across devices)
+        pp_degree: Pipeline parallel degree (model sharding across pipeline stages)
+        world_size: Total number of processes in the distributed training job
+        dp_rank: Current data parallel rank (0 to dp_degree-1)
+        tp_rank: Current tensor parallel rank (0 to tp_degree-1)
+        device: Device identifier (e.g., "cuda:0", "cuda:1")
+
+    Example:
+        >>> info = await trainer.get_info()
+        >>> p = info.parallelism
+        >>> print(f"DP={p.dp_degree}, TP={p.tp_degree}, PP={p.pp_degree}")
+        >>> print(f"Running on {p.device} (DP rank {p.dp_rank})")
+    """
+
+    dp_degree: int
+    tp_degree: int
+    pp_degree: int
+    world_size: int
+    dp_rank: int
+    tp_rank: int
+    device: str
+
+
+@dataclass
+class TrainerInfo:
+    """Static trainer and model metadata.
+
+    This contains information about the trainer configuration and model architecture
+    that doesn't change during training.
+
+    Attributes:
+        model_name: Name or path of the model being trained
+        step: Current training step
+        model_config: Model architecture configuration. Common keys include:
+            - vocab_size: int - Size of the vocabulary
+            - hidden_size: int - Hidden dimension size
+            - num_layers: int - Number of transformer layers
+            - num_attention_heads: int - Number of attention heads
+            - max_seq_len: int - Maximum sequence length
+        parallelism: Parallelism configuration for distributed training
+
+    Example:
+        >>> info = await trainer.get_info()
+        >>> print(f"Training {info.model_name} at step {info.step}")
```

Comment:
These print statements aren't actually helpful b/c they require you to run it yourself (which you can't do in the documentation). I'd maybe suggest something like:

```
>>> info = await trainer.get_config() # or info :)
>>> info.model_name
"qwen3"
>>> info.step
1
```